core/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -249,50 +249,70 @@ public CompiledScript compile(Script script, ScriptContext scriptContext) {
+        //but we check for a file script name first and an indexed script name second
+        String name = script.getScript();
 
-        if (lang == null) {
-            lang = defaultLang;
-        }
         if (logger.isTraceEnabled()) {
-            logger.trace("Compiling lang: [{}] type: [{}] script: {}", lang, script.getType(), script.getScript());
+            logger.trace("Compiling lang: [{}] type: [{}] script: {}", lang, type, name);
         }
 
         ScriptEngineService scriptEngineService = getScriptEngineServiceForLang(lang);
-        String cacheKey = getCacheKey(scriptEngineService, script.getScript());
+        String cacheKey;
+
+        CompiledScript compiledScript;
 
         if (script.getType() == ScriptType.FILE) {
-            CompiledScript compiled = staticCache.get(cacheKey); //On disk scripts will be loaded into the staticCache by the listener
-            if (compiled == null) {
-                throw new IllegalArgumentException("Unable to find on disk script " + script.getScript());
+            cacheKey = getCacheKey(scriptEngineService, name, null);
+            //On disk scripts will be loaded into the staticCache by the listener
+            compiledScript = staticCache.get(cacheKey);
+
+            if (compiledScript == null) {
+                compiledScript = new CompiledScript(type, name, lang, null);
+                throw new IllegalArgumentException("Unable to find on disk " + compiledScript);
             }
-            return compiled;
+
+            return compiledScript;
         }
 
+        //script.getScript() will be code if the script type is inline
         String code = script.getScript();
 
         if (script.getType() == ScriptType.INDEXED) {
-            final IndexedScript indexedScript = new IndexedScript(lang, script.getScript());
+            final IndexedScript indexedScript = new IndexedScript(lang, name);
+            name = indexedScript.id;
             code = getScriptFromIndex(indexedScript.lang, indexedScript.id);
-            cacheKey = getCacheKey(scriptEngineService, code);
         }
 
-        CompiledScript compiled = cache.getIfPresent(cacheKey);
-        if (compiled == null) {
-            //Either an un-cached inline script or an indexed script
-            compiled = new CompiledScript(lang, scriptEngineService.compile(code));
+        //if the script type is inline the name will be the same as the code for identification in exceptions
+        cacheKey = getCacheKey(scriptEngineService, name, code);
+        compiledScript = cache.getIfPresent(cacheKey);
+
+        if (compiledScript == null) {
+            //Either an un-cached inline script or indexed script
+            try {
+                compiledScript = new CompiledScript(type, name, lang, scriptEngineService.compile(code));
+            } catch (Exception exception) {
+                compiledScript = new CompiledScript(type, name, lang, null);
```

Comment:
Can we just print out the pieces type/name/lang here without creating a compiled script just to do a toString()?